### DISCUSSION TITLE
Docker Container Action - How to persist files in workspace

### DISCUSSION THREAD

**Discussion by velsietis** [community/community#26963] [Actions]
<p>Hi,</p>
<p>Can anybody help explain what I’m missing here?</p>
<p>I’m following <a href="https://help.github.com/en/actions/building-actions/creating-a-docker-container-action" rel="nofollow noopener">this guide</a> to run a docker container action, and trying to copy out the files.</p>
<p>It’s a pretty simple container action that runs a shell script like this:</p>
<pre><code># entrypoint.sh
cd $GITHUB_WORKSPACE
touch workspace.file
mkdir "workspace-folder"
ls -la
</code></pre>
<p>From the logs I see that both <em>workspace.file</em> and <em>workspace-folder</em> are created in the running container.</p>
<p>However, if I list the files using  <strong>ls</strong>  as the next step after the container, it seems that the <em>workspace-folder</em> has been persisted but <em>workspace-file</em> has not (after the container has exited)</p>
<pre><code># /.github/workflows/main.yml (excerpt)

steps:
      - uses: actions/checkout@v2
      - uses: ./.github/actions/my-action
      - run: ls -la $GITHUB_WORKSPACE
</code></pre>

**Comment by Yanjingzhu** [comment, marked as answer]
<p>When using a docker container action, there are some volumes configured in docker run command. </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/1/1f356e45ebe46c32786f67f0552a0eef27f3a539.png" data-download-href="/uploads/short-url/4s5h8F8hoGAWhD6SStshx9hi3RT.png?dl=1" title="docker container.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181100981-dff95c15-49b8-4ba5-af77-d43d5745e0d6.png" alt="docker container.png" data-base62-sha1="4s5h8F8hoGAWhD6SStshx9hi3RT" width="690" height="224" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/1/1f356e45ebe46c32786f67f0552a0eef27f3a539_2_690x224.png, https://aws1.discourse-cdn.com/github/optimized/2X/1/1f356e45ebe46c32786f67f0552a0eef27f3a539_2_1035x336.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/2X/1/1f356e45ebe46c32786f67f0552a0eef27f3a539_2_1380x448.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/1/1f356e45ebe46c32786f67f0552a0eef27f3a539_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">docker container.png</span><span class="informations">1509×492 44.1 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>Files inside container under /github/workspace are mapped to runner workspace /home/runner/work/my-repo-name/my-repo-name . And files inside container under /github/home are mapped to /home/runner/work/_temp/_github_home in runner machine. </p>
<p>The files you created in entrypoint.sh will be persisted after the container has exited but the empty folder will not. </p>
<p>In your workflow yml file upload artifacts actions, the path you specifed is wrong. </p>
<p>$GITHUB_WORKSPACE and $HOME is the right syntax in bash script , it could not be replaced in upload artifacts path input variable.  And $HOME in runner is point to /home/runner , it is not mapped to /github/home. You need to use /home/runner/work/_temp/_github_home instead. </p>
<p>Please try my workflow file sample:</p>
<pre><code>jobs:
  build:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: ./.github/actions/my-action
      - run: ls -la $GITHUB_WORKSPACE
      - run: ls -la /home/runner/work/_temp/_github_home
      - uses: actions/upload-artifact@v2
        with:
          name: workspace_artifacts
          path: ${{github.workspace}}
      - uses: actions/upload-artifact@v2
        with:
          name: home-artifacts
          path: /home/runner/work/_temp/_github_home</code></pre>

**Comment by velsietis** [comment]
<p>Thanks so much!  Can’t believe I didn’t think to look at that in the logs</p>
